cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java

```diff
@@ -78,38 +89,63 @@
   // caches request key to topic
   private final Map<String, String> requestKeyToLocalTopic;
 
-  private final MessagePublisher messagePublisher;
   private final long pollTimeMillis;
   private final long gracefulShutdownMillis;
   private final List<MonitorMessage> lastProgramStateMessages;
+  private final DatasetFramework dsFramework;
+  private final MultiThreadMessagingContext messagingContext;
+  private final Transactional transactional;
+  private final RetryStrategy retryStrategy;
+  private AppMetadataStore store;
```

Comment:
Why is it a field? The store can only be used inside transactional. You should pass it as parameter.